lint/capability_fields_analyser.go

```diff
@@ -0,0 +1,140 @@
+/*
+ * Cadence-lint - The Cadence linter
+ *
+ * Copyright 2019-2023 Dapper Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package lint
+
+import (
+	"github.com/onflow/cadence/runtime/ast"
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/cadence/tools/analysis"
+)
+
+func DetectCapabilityType(typeToCheck ast.Type, structWithPubCapabilities map[string]struct{}) bool {
+	const capabilityTypeName = "Capability"
+	switch downcastedType := typeToCheck.(type) {
+	case *ast.NominalType:
+		_, found := structWithPubCapabilities[downcastedType.Identifier.Identifier]
+		return downcastedType.Identifier.Identifier == capabilityTypeName || found
+	case *ast.OptionalType:
+		return DetectCapabilityType(downcastedType.Type, structWithPubCapabilities)
+	case *ast.VariableSizedType:
+		return DetectCapabilityType(downcastedType.Type, structWithPubCapabilities)
+	case *ast.ConstantSizedType:
+		return DetectCapabilityType(downcastedType.Type, structWithPubCapabilities)
+	case *ast.DictionaryType:
+		return DetectCapabilityType(downcastedType.KeyType, structWithPubCapabilities) || DetectCapabilityType(downcastedType.ValueType, structWithPubCapabilities)
+	case *ast.FunctionType:
+		return false
+	case *ast.ReferenceType:
+		return DetectCapabilityType(downcastedType.Type, structWithPubCapabilities)
+	case *ast.RestrictedType:
+		return false
+	case *ast.InstantiationType:
+		return DetectCapabilityType(downcastedType.Type, structWithPubCapabilities)
+	default:
+		panic("Unknown type")
+	}
+}
+
+func CollectStructsWithPublicCapabilities(inspector *ast.Inspector) (map[string]struct{}, map[ast.Identifier]struct{}) {
+	structWithPubCapabilities := make(map[string]struct{})
+	fieldsInStruct := make(map[ast.Identifier]struct{})
+	inspector.Preorder(
+		[]ast.Element{(*ast.CompositeDeclaration)(nil)},
+		func(element ast.Element) {
+			switch declaration := element.(type) {
+			case *ast.CompositeDeclaration:
+				{
+					if declaration.CompositeKind != common.CompositeKindStructure {
+						return
+					}
+					for _, d := range declaration.Members.Declarations() {
+						field, ok := d.(*ast.FieldDeclaration)
+						if !ok {
+							return
+						}
+						if field.Access != ast.AccessPublic {
+							return
+						}
+						if DetectCapabilityType(field.TypeAnnotation.Type, structWithPubCapabilities) {
```

Comment:
Currently, `DetectCapabilityType` detects a type which is or includes the `Capability` type purely on the syntax tree (`ast.Type`). Maybe consider operating on the semantic level, i.e. `sema.Type`.

Reply:
The same can be done for `sema.Type`, e.g. there's `sema.ConstantSizedType`, `sema.DictionaryType`, etc.

Reply:
I tried to take a look today. I see that usually ast is used when checks are simple like in reference_operator_analyser.go etc... I found in redundant_cast_analyser.go sema is used. 
The conversion from ast to sema is done like 
```redundantType := elaboration.StaticCastTypes(castingExpression)```
where ```castingExpression``` is of ```*ast.CastingExpression``` type. 

So, I went to elaboration.go but  couldn't find any function like StaticCastTypes that would take ```*ast.FieldDeclaration``` as an argument.  And I'd need to elaborate on a declaration.
Do you think here I an easy way to get from ```*ast.FieldDeclaration``` to relevant ```sema.Type```s ?

Reply:
Hi @turbolent , so what do you think about my previous comment ? I couldn't find an existing function in elaboration that would allow me to convert so sema.

Reply:
Sorry for dropping the ball here Boris.

It's OK to leave it as-is for now. Given there's no shadowing allowed fore types, it should be sufficient to detect `Capability` purely syntactically (i.e. through the AST instead of sema).

Reply:
We might have to do this using the sema-type, because the contract could be using types that are imported, and for them, the ast-node (`ast.Type`) might not be available at this stage.
e.g:
```cadence
import ImprtedContract from 0x01

pub contract MyContract {

    // Assume `ImprtedContract.ContractData` has a public capability field.
    // It won't be able to report an error here only by looking at the syntax-tree
    pub var contractData: ImprtedContract.ContractData
```